Sources/BraveUI/SwiftUI/WebImageReader.swift

```diff
@@ -5,81 +5,82 @@
 
 import SwiftUI
 import SDWebImage
-import SDWebImageSVGNativeCoder
+import BraveCore
 
-private class WalletWebImageManager: ObservableObject {
-  /// loaded image, note when progressive loading, this will published multiple times with different partial image
-  @Published public var image: UIImage?
-  /// loaded image data, may be nil if hit from memory cache. This will only published once loading is finished
-  @Published public var isFinished: Bool = false
-  /// loading error
-  @Published public var error: Error?
-
-  private var manager = SDWebImageManager.shared
-  private var operation: SDWebImageOperation?
+public protocol WebImageDownloaderType: AnyObject {
+  func downloadImage(url: URL?) async -> UIImage?
+  func imageFromData(data: Data) -> UIImage?
+}
 
-  private var supportedCoders: [SDImageCoder] = [SDImageSVGNativeCoder.shared, SDImageAPNGCoder.shared, SDImageGIFCoder.shared]
-  
-  init() {}
+struct WebImageDownloaderKey: EnvironmentKey {
+  static var defaultValue: WebImageDownloaderType? = SDWebImageManager.shared
+}
 
-  func load(url: URL?, options: SDWebImageOptions = []) {
-    operation = manager.loadImage(
-      with: url, options: options, progress: nil,
-      completed: { [weak self] image, data, error, _, finished, _ in
-        guard let self = self else { return }
-        self.image = image
-        self.error = error
-        if finished {
-          self.isFinished = true
-        }
-      })
+extension EnvironmentValues {
+  public var webImageDownloader: WebImageDownloaderType? {
+    get { self[WebImageDownloaderKey.self] }
+    set { self[WebImageDownloaderKey.self] = newValue }
   }
+}
 
-  func cancel() {
-    operation?.cancel()
-    operation = nil
+extension SDWebImageManager: WebImageDownloaderType {
+  @MainActor public func downloadImage(url: URL?) async -> UIImage? {
+    return await withCheckedContinuation { continuation in
+      loadImage(with: url, progress: nil) { image, _, _, _, _, _ in
+        continuation.resume(returning: image)
+      }
+    }
```

Comment:
This should become a `withTaskCancellationHandler` that holds onto the `SDWebImageCombinedOperation` and calls cancel on it when the task is cancelled:

```swift
@MainActor public func downloadImage(url: URL?) async -> UIImage? {
  var operation: SDWebImageCombinedOperation?
  return await withTaskCancellationHandler {
    operation?.cancel()
  } operation: {
    await withCheckedContinuation { continuation in
      operation = loadImage(with: url, progress: nil) { image, _, _, _, _, _ in
        continuation.resume(returning: image)
      }
    }
  }
}
```